### DISCUSSION TITLE
How to use the tools?

### DISCUSSION THREAD

**Discussion by mkarolczyk** [open-webui/open-webui#3746] [Q&A]
I have prepared a python script that communicates with Todoist. I select tools on the chat page, but no model sees the option to use tools. Tools provided by the community do not work either
What am I doing wrong?

**Comment by justinh-rahb** [comment]
We couldn't possibly weigh in on this without 1) the tool code in question and 2) some logs of what errors are occurring. That said, we're not going to be able to sustain offering support for third-party code forever, though we're willing to help out _right now_ with early adopters to get the wrinkles ironed out.

Edit: also, you need to use a model that supports tool-use. Right now the best to use is obviously gpt-4o, and its not even close. The _best_ open model right now is Llama3 8B or Mistral 7B v0.3, and that's honestly not saying a lot.. they can handle some basic tools with a small number of parameters, but anything that involves huge amounts of data will go over Llama3's context limit pretty quickly for example.

**Reply by mkarolczyk** [reply]
I have tried on both the Ollama and gpt-4o models, each time the model does not see the tools.  I've added myself the code from the OpenWebUi page about downloading data from wikipedia: https://openwebui.com/t/pyotrgrowpotkin/wikidata/ . The gpt-4o says it cannot download the data directly.

My question is this - does something extra need to be done to use tools besides adding the tools code and then activating a specific tool in the chat window?

![CleanShot 2024-07-10 at 17 06 40@2x](https://github.com/open-webui/open-webui/assets/2911317/2b0d256a-bfb1-4f85-8092-0599af02947d)

**Reply by david02871** [reply]
I have the same issue, also using GPT4-o and custom tools. The tool is being run but gpt4-o is insisting that it can not access the web and ignoring the tool responses.

**Comment by adamhj** [comment]
~~I have the same issue, open-webui 0.5.10 installed from pip. Through network packet capture, I can see that "available tools" is empty
![open-webui_tool_bug](https://github.com/user-attachments/assets/5dd95e94-1809-4d1d-bfe9-72888ac20111)~~

**Reply by adamhj** [reply]
Sorry, it's my mistake, I've just find  that the tool script I've installed is buggy. Switch to another tool  works fine.
